src/core/vectortile/qgsvectortileloader.cpp

```diff
@@ -0,0 +1,265 @@
+  qgsvectortileloader.cpp
+  --------------------------------------
+  Date                 : March 2020
+  Copyright            : (C) 2020 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsvectortileloader.h"
+
+#include <QtDebug>
+#include <QEventLoop>
+
+#include <zlib.h>
+
+#include "qgsblockingnetworkrequest.h"
+#include "qgsmbtilesreader.h"
+#include "qgsnetworkaccessmanager.h"
+#include "qgsvectortileutils.h"
+
+QgsVectorTileLoader::QgsVectorTileLoader( const QString &uri, int zoomLevel, const QgsTileRange &range, const QPointF &viewCenter, QgsFeedback *feedback )
+  : mEventLoop( new QEventLoop )
+  , mFeedback( feedback )
+{
+  if ( feedback )
+  {
+    connect( feedback, &QgsFeedback::canceled, this, &QgsVectorTileLoader::canceled, Qt::QueuedConnection );
+
+    // rendering could have been canceled before we started to listen to canceled() signal
+    // so let's check before doing the download and maybe quit prematurely
+    if ( feedback->isCanceled() )
+      return;
+  }
+
+  qDebug() << "starting loader";
+  QVector<QgsTileXYZ> tiles = QgsVectorTileUtils::tilesInRange( range, zoomLevel );
+  QgsVectorTileUtils::sortTilesByDistanceFromCenter( tiles, viewCenter );
+  for ( QgsTileXYZ id : qgis::as_const( tiles ) )
+  {
+    loadFromNetworkAsync( id, uri );
+  }
+}
+
+QgsVectorTileLoader::~QgsVectorTileLoader()
+{
+  qDebug() << "terminating loader";
+
+  if ( !mReplies.isEmpty() )
+  {
+    // this can happen when the loader is terminated without getting requests finalized
+    // (e.g. downloadBlocking() was not called)
+    canceled();
+  }
+}
+
+void QgsVectorTileLoader::downloadBlocking()
+{
+  qDebug() << "starting event loop" << mReplies.count() << "requests";
+
+  if ( mFeedback && mFeedback->isCanceled() )
+  {
+    qDebug() << "actually not - we were cancelled";
+    return; // nothing to do
+  }
+
+  mEventLoop->exec( QEventLoop::ExcludeUserInputEvents );
+
+  qDebug() << "download blocking finished";
+
+  Q_ASSERT( mReplies.isEmpty() );
+}
+
+void QgsVectorTileLoader::loadFromNetworkAsync( const QgsTileXYZ &id, const QString &requestUrl )
+{
+  QString url = QgsVectorTileUtils::formatXYZUrlTemplate( requestUrl, id );
+  QNetworkRequest request( url );
+  // TODO: some extra headers? QgsSetRequestInitiatorClass / auth / "Accept" header
```

Comment:
There's a macro you should call after creating requests - see elsewhere. It adds metadata to the request used for the network logger